server/controllers/auth.js

```diff
@@ -54,36 +56,53 @@ function login(req, res, next) {
     project_permission JOIN project ON user.id = project_permission.user_id 
     AND project.id = project_permission.project_id
     WHERE 
-      user.username = ? AND user.password = PASSWORD(?) 
-      AND project_permission.project_id = ?;
+      user.username = ? AND project_permission.project_id = ?;
   `;
 
   const sqlUser = `
-    SELECT user.id FROM user
-    WHERE user.username = ? AND user.password = PASSWORD(?)`;
+    SELECT user.id, user.password FROM user
+    WHERE user.username = ?`;
 
   const sqlPermission =
     `SELECT permission.id FROM permission
     JOIN user ON user.id = permission.user_id
-    WHERE user.username = ? AND user.password = PASSWORD(?)`;
+    WHERE user.username = ?`;
 
 
-  db.exec(sql, [username, password, projectId])
+  db.exec(sql, [username, projectId])
     .then((rows) => {
       param.connect = rows;
-
-      return db.exec(sqlUser, [username, password]);
+      return db.exec(sqlUser, [username]);
     })
     .then((rows) => {
       param.user = rows;
-      return db.exec(sqlPermission, [username, password]);
+      return db.exec(sqlPermission, [username]);
     })
     .then((rows) => {
       param.permission = rows;
+
+      // check if username is correct
+      if (param.user.length === 0) {
+        throw new Unauthorized('The user name provided is bad.');
+      }
+
+      // check if password is set
+      if (!password) {
+        throw new Unauthorized('No password provided.');
+      }
+
+      return util.checkString(password, param.user[0].password);
+    })
+    .then((resp) => {
       const connect = param.connect.length;
       const permission = param.permission.length;
-      const user = param.user.length;
 
+      // Will be false if the password does not much
+      if (!resp) {
+        throw new Unauthorized('Bad password, Please Try again', 'FORM.ERRORS.PASSWORD_MATCH');
```

Comment:
Hmm.. I can see the usability argument.

In my view, the reason we want to use `bcrypt` over `PASSWORD()` is to start getting serious about enterprise security, in case we ever actually do have to conform to something like HIPPA requirements.  So, in my view, we should try and adopt security best practices.

According to [The Open Web Application Security Project](https://www.owasp.org/index.php/Authentication_Cheat_Sheet), you should always give a generic error message.  Here's the quote:
> An application should respond with a generic error message regardless of whether the user ID or password was incorrect. It should also give no indication to the status of an existing account.

For this reason, I'm a fan of using a uniform authentication failure message.  However, I can be swayed if our users can't figure it out.